### PR TITLE
Don't park immediately, as unpark is very expensive

### PR DESCRIPTION
**Pull request by stuartwdouglas**
Instead we spin/yield for a bit, to see if we get another task

### CODE REVIEW

File: src/main/java/org/jboss/threads/EnhancedQueueExecutor.java [hunk 1]
```diff
@@ -149,6 +150,10 @@
      */
     static final boolean DISABLE_MBEAN = readBooleanPropertyPrefixed("register-mbean", readProperty("org.graalvm.nativeimage.imagecode", null) != null);
 
+    /**
+     * The number of times a thread should spin/yield before actually parking.
+     */
+    static final int SPINS = 1 << 7;
```
[the remark this answers is not in the document]
Comment (stuartwdouglas) [reply src/main/java/org/jboss/threads/EnhancedQueueExecutor.java]: This was based on looking at the code in LinkedTransferQueue. It uses this value for the first thread node, and 1 <<6 if there is another thread already spinning. I just went for a simplified approach here.

The yield logic with the random number is also from there. I tried some different approaches but the ones from linked transfer queue seemed to give the best performance.

File: src/main/java/org/jboss/threads/EnhancedQueueExecutor.java [hunk 2]
```diff
@@ -2148,6 +2160,62 @@ Runnable getTask() {
         PoolThreadNode getNext() {
             return (PoolThreadNode) super.getNext();
         }
+
+        void park(EnhancedQueueExecutor enhancedQueueExecutor) {
+            int spins = SPINS;
+            ThreadLocalRandom tl = ThreadLocalRandom.current();
+            while (spins > 0) {
+                if (tl.nextInt(SPINS) == 0) {
+                    Thread.yield();
```
[the remark this answers is not in the document]
Comment (stuartwdouglas) [reply src/main/java/org/jboss/threads/EnhancedQueueExecutor.java]: I just copied from LinkedTransferQueue. The big problem here is that to do this properly we need a comprehensive benchmark suite that covers lots of different use cases, and a benchmark environment that lets us test it in different environments.

I basically just assumed that the approach taken in LinkedTransferQueue had a lot of thought put into it, and copied it, as I don't think we are going to have the resources to try and figure out an optimal approach any time soon.

**Comment by dmlloyd** [comment]
It looks like the theory here is that a thread should avoid going to sleep waiting for a task if a task is likely to appear soon, that is, if we're right at the cusp of keeping all the threads busy but beyond into queueing territory?  Whereas @carterkozak's benchmarks mainly cover the case where the queue has plenty of available warm threads to run tasks?

**Comment by stuartwdouglas** [comment]
The theory is that if a thread has jump completed a task then it is better for it to spin for a little bit to see if another task appears soon, which it almost certainly will on busy systems. This is really for situations when the queue is generally empty or close to it.

The park/unpark process seems to be relatively expensive, even for the thread that calls unpark, so for systems that use an IO thread to dispatch to a worker the IO thread will end up spending a lot of time in 'unpark', which has a massive impact on the overall performance.

**Comment by carterkozak** [comment]
I added a couple lines to my benchmark to profile invocations of unpark. Using a build from master it's invoked 98 million times, where this patch results in a single call thus the 50% speedup.

**Comment by FroMage** [comment]
As I said on zulip, this doesn't appear to change anything on TFB plaintext: we're still at 40% CPU utilisation.

**Comment by dmlloyd** [comment]
What SHA did you test with @FroMage ?

**Comment by FroMage** [comment]
8300beb029c8b188f25bfe5968560b48691a8f16

**Pull request merged**